src/Csv/CsvGenerator.fs

```diff
@@ -15,79 +15,121 @@ open ProviderImplementation.QuotationBuilder
-      let rowVarExpr = Expr.Var rowVar
-      let body = Expr.NewArray(typeof<string>, [ for field in fields -> field.ConvertBack rowVarExpr ])
-      let delegateType = typedefof<Func<_,_>>.MakeGenericType(rowErasedType, typeof<string[]>)
-
-      Expr.NewDelegate(delegateType, [rowVar], body)
-
-    csvType, csvErasedTypeWithRowErasedType, rowType, stringArrayToRow, rowToStringArray
+    type private FieldInfo =
+        {
+            /// The representation type that is part of the tuple we extract the field from
+            TypeForTuple: Type
+            /// The provided property corresponding to the field
+            ProvidedProperty: ProvidedProperty
+            Convert: Expr -> Expr
+            ConvertBack: Expr -> Expr
+            /// The provided parameter corresponding to the field
+            ProvidedParameter: ProvidedParameter
+        }
+
+    let generateTypes asm ns typeName (missingValuesStr, cultureStr) inferredFields =
+
+        let fields =
+            inferredFields
+            |> List.mapi (fun index field ->
+                let typ, typWithoutMeasure, conv, convBack =
+                    ConversionsGenerator.convertStringValue missingValuesStr cultureStr field
+
+                let propertyName = NameUtils.capitalizeFirstLetter field.Name
+
+                { TypeForTuple = typWithoutMeasure
+                  ProvidedProperty =
+                    ProvidedProperty(
+                        propertyName,
+                        typ,
+                        getterCode =
+                            fun (Singleton row) ->
+                                match inferredFields with
+                                | [ _ ] -> row
+                                | _ -> Expr.TupleGet(row, index)
+                    )
+                  Convert = fun rowVarExpr -> conv <@ TextConversions.AsString((%%rowVarExpr: string[]).[index]) @>
+                  ConvertBack =
+                    fun rowVarExpr ->
+                        convBack (
+                            match inferredFields with
+                            | [ _ ] -> rowVarExpr
+                            | _ -> Expr.TupleGet(rowVarExpr, index)
+                        )
+                  ProvidedParameter = ProvidedParameter(NameUtils.niceCamelName propertyName, typ) })
+
+        // The erased row type will be a tuple of all the field types (without the units of measure).  If there is a single column then it is just the column type.
+        let rowErasedType =
+            match fields with
+            | [ field ] -> field.TypeForTuple
+            | _ -> FSharpType.MakeTupleType([| for field in fields -> field.TypeForTuple |])
+
+        let rowType =
+            ProvidedTypeDefinition("Row", Some rowErasedType, hideObjectMethods = true, nonNullable = true)
+
+        let ctor =
+            ProvidedConstructor(
+                [ for field in fields -> field.ProvidedParameter ],
+                invokeCode =
```

Comment:
These `invokeCode` expand out a fair bit, I will pre-adjust them to be inner function bindings